rfcs/0030-ensure-index-less-than-length-in-since/0030-ensure-index-less-than-length-in-since.md

```diff
@@ -1 +1,87 @@
+
+```
+E > epoch
+or (E == epoch and I / L >= index/length)
+```
+
+This is not consistent when using relative flag in the `since` field, where the `value` is considered as a rational number and added to the cell start epoch.
+
+This document suggests adding a new rule to verify that when `since` uses epoch as the unit, it must ensure that index is less than the length.
+
+See RFC17 [Transaction valid since](../0017-tx-valid-since/0017-tx-valid-since.md) for details of the `since` field encoding patterns.
+
+## Specification
+
+### Since Absolute Epoch
+
+When an input `since` field is present, and
+
+* The `metric_flag` is epoch (01).
+* The `relative_flag` is absolute (0).
+
+Providing that `value` is the encoded number `epoch + index / length`, the transaction is valid only if `index` is less than `length` or both `index` and `length` are zero \*.
+
+> \* When both `index` and `length` are zero, `value` is converted to `epoch + 0 / 1`. 
+
+If the current epoch number with the fraction is `E + I / L`, the transaction is considered mature when
+
+```
+E + I / L >= epoch + index / length
+```
+
+using rational number operations.
+
+### Since Relative Epoch
+
+When an input `since` field is present, and
+
+* The `metric_flag` is epoch (01).
+* The `relative_flag` is relative (1).
+
+Providing that `value` is the encoded number `epoch + index / length`, the transaction is valid only if `index` is less than `length` or both `index` and `length` are zero \*.
+
+> \* When both `index` and `length` are zero, `value` is converted to `epoch + 0 / 1`. 
+
+If the cell creation epoch number with fraction is `e + i / l`, and the current epoch number with the fraction is `E + I / L`, the transaction is considered mature when
+
+```
+E + I / L >= e + i / l + epoch + index / length
+```
+
+using rational number operations.
+
+## Test Vectors
+
+When `since` uses the absolute epoch `99 + 360 / 180`, and the current epoch is `100 + 0 / 180`, the transaction is mature using the old consensus rule but is immature using the new rule.
+
+## Deployment
+
+The deployment can be performed in two stages.
+
+The first stage will activate the new consensus rule starting from a specific epoch. The mainnet and testnet will use different starting epochs and all other chains will use the new rule from epoch 0.
+
+The second stage is optional. After the new rule has been activated, and the blocks in the chain before activation can also pass the new consensus rule, the old rule can be removed and all chains will use the new rule starting from the genesis block.
+
+## Backward compatibility
+
+The new rule is striker than the old one thus it can be deployed via a soft fork. When most nodes have upgraded to the new version, the old versions can keep up to date although blocks generated by old versions may be rejected by new version nodes. 
```

Comment:
```suggestion
The new rule is stricter than the old one thus it can be deployed via a soft fork. When most mining nodes have upgraded to the new version, the old version full nodes can keep up to date. Blocks generated by old version mining nodes may be rejected by new version full nodes. 
```